File: python/basics/arrays/238_product_array_except_self.py
'''
https://leetcode.com/problems/product-of-array-except-self/description/?envType=study-plan-v2&envId=top-interview-150


'''

def product_of_array_except_self_bruteforce(nums: list[int])->list[int]:
    n = len(nums)
    result = [1]*n
    
    for i in range(n):
        for j in range(n):
            if i != j:
                result[i] = result[i] * nums[j]
    return result

#time: O(n)
#space: O(n)
'''
create pre and post list, multiply these
'''
def product_of_array_except_self(nums: list[int])->list[int]:
    n = len(nums)
    result = [0]*n
    pre = [0]*n
    pre[0]=1
    post = [0]*n
    post[n-1]=1
    
    for i in range(1,n):
        pre[i] = pre[i-1] * nums[i-1]
    print("pre: " + str(pre))
    for i in range(n-2,-1,-1):
        post[i] = post[i+1] * nums[i+1]
    print("post: " + str(post))

    for i in range(n):
        result[i]=pre[i] * post[i]
    return result

'''
only create the pre list
'''
def product_of_array_except_self_space(nums: list[int])->list[int]:
    n = len(nums)
    pre = [0]*n
    pre[0]=1
    
    for i in range(1,n):
        pre[i] = pre[i-1] * nums[i-1]
    print("pre: " + str(pre))
    mult = 1
    for i in range(n-1, -1, -1):
        pre[i]=pre[i] * mult
        mult = mult * nums[i]
    return pre


input = [1,2,3,4]
result = product_of_array_except_self(input)
print("result:" + str(result))

input = [-1,1,0,-3,3]
result = product_of_array_except_self(input)
print("result:" + str(result))


input = [1,2,3,4]
result = product_of_array_except_self_bruteforce(input)
print("result brute:" + str(result))

input = [-1,1,0,-3,3]
result = product_of_array_except_self_bruteforce(input)
print("result brute:" + str(result))


input = [1,2,3,4]
result = product_of_array_except_self_space(input)
print("result space:" + str(result))

input = [-1,1,0,-3,3]
result = product_of_array_except_self_space(input)
print("result space:" + str(result))

File: problem_solving/arrays/cpp/238_product_array_except_self.cpp
/*
'''
https://leetcode.com/problems/product-of-array-except-self/description/?envType=study-plan-v2&envId=top-interview-150

Approach I:
    using pre and post arrays.

Approach II:
    using only one array

'''
*/

#include "../../../cpp/common_header.h"
#include "../../../cpp/utils/utils.h"
//todo


std::vector<int>  product_of_array_except_self(const std::vector<int>& nums) {
    int len = nums.size();
    std::vector<int> result(len, 1);
    for(int i = 1; i < len; ++i) {
        result[i] = result[i-1] * nums[i-1];
    }
    int mult = 1;
    for(int i = len-2; i >= 0; --i) {
        mult *= nums[i+1];
        result[i] *= mult;
    }
    return result;
}

int main() {
    std::vector<int> input;
    input = {1,2,3,4};
    std::vector<int> result = product_of_array_except_self(input);
    print_vector(result);

    input = {-1,1,0,-3,3};
    result = product_of_array_except_self(input);
    print_vector(result);
    return 0;
}